<a href="https://colab.research.google.com/github/ayannj13/Data-Science/blob/main/Homework4_AyanJamal.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [1]:
import pandas as pd
url = "https://raw.githubusercontent.com/ogut77/DataScience/master/insurance.csv"
df = pd.read_csv(url)


In [2]:
df.head()

age     sex     bmi  children smoker     region      charges
0   19  female  27.900         0    yes  southwest  16884.92400
1   18    male  33.770         1     no  southeast   1725.55230
2   28    male  33.000         3     no  southeast   4449.46200
3   33    male  22.705         0     no  northwest  21984.47061
4   32    male  28.880         0     no  northwest   3866.85520

Context in Insurance Data
This dataset is often used to predict charges based on the other variables (age, sex, bmi, children, smoker, region). For example:

Input Variables (X): age, sex, bmi, children, smoker, region (features used to make predictions).

Output Variable (y): charges (what you’re trying to predict).

Describtion of variables
1. Age
Description: The age of the individual (the insured person).
Type: Numerical (integer).
Example Values: 19, 45, 62, etc.
Role in Insurance: Age is a key factor in determining insurance charges. Older individuals often have higher medical costs (and thus higher charges) due to increased health risks.
2. Sex
Description: The gender of the individual.
Type: Categorical (text or binary).
Example Values: "male," "female"
Role in Insurance: Gender can influence insurance charges because health risks and medical expenses may differ between males and females (e.g., pregnancy-related costs for females).
3. BMI (Body Mass Index)
Description: A measure of body fat based on height and weight (calculated as weight in kg divided by height in meters squared).
Type: Numerical (float).
Example Values: 25.3, 30.1, 18.5, etc.
Role in Insurance: Higher BMI often correlates with increased health risks (e.g., obesity-related conditions like diabetes or heart disease), leading to higher insurance charges.
4. Children
Description: The number of children (dependents) covered under the individual’s insurance plan.
Type: Numerical (integer).
Example Values: 0, 1, 3, etc.
Role in Insurance: More children can increase insurance costs slightly, as it may reflect additional healthcare needs, though the effect is often less pronounced than other factors like smoking or age.
5. Smoker
Description: Indicates whether the individual smokes tobacco.
Type: Categorical (text or binary).
Example Values: "yes," "no" .
Role in Insurance: Smoking is a major factor in insurance charges. Smokers typically have much higher medical costs due to risks like lung disease or cancer, so their charges are significantly elevated.
6. Region
Description: The geographic region where the individual lives.
Type: Categorical (text).
Example Values: "northeast," "southeast," "southwest," "northwest" (common in U.S.-based datasets).
Role in Insurance: Charges can vary by region due to differences in healthcare costs, lifestyle factors, or local insurance regulations.
7. Charges
Description: The insurance charges (or premiums/costs) billed to the individual, typically in a currency like USD.
Type: Numerical (float).
Example Values: 1684.52, 11234.89, 32050.23, etc.
Role in Insurance: This is usually the target variable (output) in predictive modeling. It represents the amount the insurance company charges, influenced by all the other columns (age, sex, BMI, etc.).



In [3]:
#1. Check if there is null value in dataset df (5 pt)

print(df.isnull().sum())

age         0
sex         0
bmi         0
children    0
smoker      0
region      0
charges     0
dtype: int64


In [4]:
#2. Assign charges to y  and others to X using df. y is output variable and X is input variables (5 pt)
y = df['charges']

X = df.drop('charges', axis=1)

In [5]:
#3. Use  get_dummies() function from the pandas library to convert categorical variables in a DataFrame (X).
# Drop first drops the first category’s dummy variable to avoid multicollinearity (5 pt)

X = pd.get_dummies(X, drop_first=True)


In [6]:
#Use following methods for the evaluation on test and train data
def evalmetric(y,ypred):
 from scipy.stats import pearsonr
 import numpy as np
 e = y - ypred
 mse_f = np.mean(e**2)
 rmse_f = np.sqrt(mse_f)
 mae_f = np.mean(abs(e))
 mape_f = 100*np.mean(abs(e/y))
 crl, _ = pearsonr(y, ypred)
 r2_f = crl*crl
 print("MSE:", mse_f)
 print("RMSE:", rmse_f)
 print("MAE:",mae_f)
 print("MAPE:",mape_f)
 print("R-Squared:", round(r2_f, 4))


In [7]:
#4.Get the correlation between X variables and y variables.(5 pt)

df_corr = pd.concat([X, y], axis=1)

correlation_matrix = df_corr.corr()

print(correlation_matrix['charges'].sort_values(ascending=False))

charges             1.000000
smoker_yes          0.787251
age                 0.299008
bmi                 0.198341
region_southeast    0.073982
children            0.067998
sex_male            0.057292
region_northwest   -0.039905
region_southwest   -0.043210
Name: charges, dtype: float64


In [8]:
#5.Split a dataset into 25%  of data as test data  and 75% of data as training data ( pt)

from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.25, random_state=42)


In [ ]:
#6. Using Decision Tree and Linear Regression methods, compare the performance results on both test and training data
#to determine which one is more likely to overfit and which is more likely to underfit.
# Do you think that Lasso and Ridge regularization are more likely to improve the results of Linear model test data,
# or would Random Forest or Boosting methods are more likely to improve the results of Decison tree test data?
#Explain your reasoning.(35 pt)

In [12]:
#Linear regression
from sklearn.linear_model import LinearRegression

#training the model
lr_model = LinearRegression()
lr_model.fit(X_train, y_train)

# predictions
y_train_pred_lr = lr_model.predict(X_train)
y_test_pred_lr = lr_model.predict(X_test)

# evaluation
print("Linear Regression - Train Data:")
evalmetric(y_train, y_train_pred_lr)

print("\nLinear Regression - Test Data:")
evalmetric(y_test, y_test_pred_lr)


Linear Regression - Train Data:
MSE: 37004502.18409475
RMSE: 6083.132596294014
MAE: 4183.153367011969
MAPE: 42.26867490005659
R-Squared: 0.745

Linear Regression - Test Data:
MSE: 35117755.73613632
RMSE: 5926.023602394469
MAE: 4243.654116653137
MAPE: 44.468185116980976
R-Squared: 0.7676


In [14]:
#decision tree
from sklearn.tree import DecisionTreeRegressor

# train the model
dt_model = DecisionTreeRegressor(random_state=42)
dt_model.fit(X_train, y_train)

# predictions
y_train_pred_dt = dt_model.predict(X_train)
y_test_pred_dt = dt_model.predict(X_test)

# evaluation
print("Decision Tree - Train Data:")
evalmetric(y_train, y_train_pred_dt)

print("\nDecision Tree - Test Data:")
evalmetric(y_test, y_test_pred_dt)


Decision Tree - Train Data:
MSE: 182648.17106092346
RMSE: 427.37357318969015
MAE: 19.084122482552342
MAPE: 0.47775094523744194
R-Squared: 0.9987

Decision Tree - Test Data:
MSE: 36580230.16400095
RMSE: 6048.159237652473
MAE: 2693.294304847761
MAPE: 29.869257985320623
R-Squared: 0.7805


In [15]:
#Answers to 6th task:

**1) Which model overfits and which underfits?**
Decision Tree is overfitting because its train R-squared is 0.9987 which is almost perfect, but test R-squared drops to 0.7805. This shows there is huge performance gap! It memorized the training data but it struggles on unseen data when it tries to make predictions.

However, Linear Regression is more balanced as its train R-squared is 0.745, and test R-squared is 0.7676. Train and test scores are close, which means the model generalizes better but it may not capture complex relationships.

**2) Which method is better to improve each model?**

**For Linear Regression → Lasso/Ridge**

Lasso and Ridge regression can help improve Linear Regression by reducing underfitting. They do this by shrinking the importance of less significant features. But, since the train and test scores for Linear Regression are already quite close, the improvement will likely be small.

**For Decision Tree → Random Forest / Boosting**

For the Decision Tree model, the best way to improve test performance is to use Random Forest or Boosting methods. Random Forest helps by combining multiple trees, and this makes the model more stable and reduces overfitting. Boosting methods, such as XGBoost, or Catboost improve accuracy by focusing on the mistakes made in previous steps. These techniques will improve the Decision Tree model more than regularization would help the Linear Regression model.


In [ ]:
#7. Explain performance of linear regressin on test data
# using  Root mean squared error, mean absolute error, mean absolute percentage error and R2 metric (10 pt)


**7th question answer:**
Linear Regression performs moderately well on test data.

The performance of Linear Regression on the test data can be explained using these metrics:

**Root Mean Squared Error (RMSE):** 5926.02
This means, the model’s predictions have an average error of about $5926. RMSE gives more weight to larger errors, so it highlights bigger mistakes.

**Mean Absolute Error (MAE):** 4243.65
This shows the average error size, meaning have an average error of about $4243.

**Mean Absolute Percentage Error (MAPE)**: 44.47%
This means the average prediction mistake is about by about 44% of the actual charges. This is quite a high error rate.

**R-Squared (R²):** 0.7676
This means the model can explain about 76.76% of the changes in insurance charges. This is a fairly good result, but it also shows that the model misses some important factors that affect the charges.

In general, Linear Regression explains a good part of the data, but its errors (RMSE, MAE, and MAPE) are quite large which shows the model is not fully accurate and it may be too simple to capture all important relationships in the data.

In [ ]:
#8. Use Random Forest and Boosting methods (XGBoost, LightGBM, and CatBoost)
#to obtain the evaluation scores on  test data.
#Which Boosting technique yielded the best performance on the test data based on the R² metric?
#Did you achieve a better result compared to Random Forest on the test data based on the R² metric?
#If there is improvement on Random forest or boosting methods over decison tree, explain  (30 pt)

In [18]:
#8th task Random forest

from sklearn.ensemble import RandomForestRegressor
rf_model = RandomForestRegressor(n_estimators=100, random_state=42)

rf_model.fit(X_train, y_train)

# predict on test data
y_test_pred_rf = rf_model.predict(X_test)

print("Random Forest - Test Data:")
evalmetric(y_test, y_test_pred_rf)


Random Forest - Test Data:
MSE: 23114410.14038345
RMSE: 4807.7448081593775
MAE: 2653.61478095801
MAPE: 30.27386314379988
R-Squared: 0.8506


In [19]:
#8th task xgboost
from xgboost import XGBRegressor
xgb_model = XGBRegressor(n_estimators=100, random_state=42)

xgb_model.fit(X_train, y_train)

# predict on test data
y_test_pred_xgb = xgb_model.predict(X_test)

print("XGBoost - Test Data:")
evalmetric(y_test, y_test_pred_xgb)


XGBoost - Test Data:
MSE: 26433443.13176504
RMSE: 5141.346431798293
MAE: 2957.213261792119
MAPE: 34.57266690344694
R-Squared: 0.8301


In [20]:
#8th task lightgbm
from lightgbm import LGBMRegressor

lgbm_model = LGBMRegressor(n_estimators=100, random_state=42)
lgbm_model.fit(X_train, y_train)

# predict on test data
y_test_pred_lgbm = lgbm_model.predict(X_test)

print("LightGBM - Test Data:")
evalmetric(y_test, y_test_pred_lgbm)


[LightGBM] [Info] Auto-choosing row-wise multi-threading, the overhead of testing was 0.001133 seconds.
You can set `force_row_wise=true` to remove the overhead.
And if memory is not enough, you can set `force_col_wise=true`.
[LightGBM] [Info] Total Bins 319
[LightGBM] [Info] Number of data points in the train set: 1003, number of used features: 8
[LightGBM] [Info] Start training from score 13267.935814
LightGBM - Test Data:
MSE: 22005117.949021608
RMSE: 4690.961303296117
MAE: 2700.720352413941
MAPE: 33.101463593940274
R-Squared: 0.8553


In [23]:
#8th task catboost
!pip install catboost


   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 98.7/98.7 MB 7.9 MB/s eta 0:00:00


In [24]:
#catboost
from catboost import CatBoostRegressor
cat_model = CatBoostRegressor(n_estimators=100, verbose=0, random_state=42)

cat_model.fit(X_train, y_train)

#predict on test data
y_test_pred_cat = cat_model.predict(X_test)

print("CatBoost - Test Data:")
evalmetric(y_test, y_test_pred_cat)


CatBoost - Test Data:
MSE: 21964464.736067303
RMSE: 4686.626157063021
MAE: 2667.445931179085
MAPE: 31.719396003541583
R-Squared: 0.855


**8th task questions' answers:**

**1) Which Boosting technique yielded the best performance on the test data based on the R² metric?**

The best Boosting method is LightGBM, because its R-squared score is 0.8553, which is the highest R-squared score out of all the boosting methods I've tried (XGBoost, LightGBM, and CatBoost).

**2) Did you achieve a better result compared to Random Forest on the test data based on the R² metric?**

Yes, LightGBM got an R-squared of 0.8553, which is slightly better than Random Forest with R-squared of 0.8506. The difference is quite small, but LightGBM still performed better.

**3) If there is improvement on Random forest or boosting methods over decison tree, explain**

Yes, all of the **ensemble methods** (Random Forest, XGBoost, LightGBM, CatBoost) did better than the single Decision Tree.
The Decision Tree had R-squared of 0.7805 on the test data, which is much lower than all the other methods.

The reason is because a single Decision Tree easily overfits to the training data. It memorizes the training set instead of learning general patterns, so it cannot generalize well on the test data.

Random Forest fixes this issue by combining many trees together, which reduces overfitting.

Boosting methods (like LightGBM, XGBoost, and CatBoost) build trees step by step, and each tree focuses on fixing the mistakes made by the tree before it. This helps them learn better patterns and generalize better to new data.
That’s why these methods performed better than the single Decision Tree.